lib/phoenix_live_view/upload_config.ex

```diff
@@ -0,0 +1,570 @@
+defmodule Phoenix.LiveView.UploadEntry do
+  @moduledoc """
+  The struct representing an upload entry.
+  """
+
+  alias Phoenix.LiveView.UploadEntry
+
+  defstruct progress: 0,
+            upload_config: nil,
+            upload_ref: nil,
+            ref: nil,
+            valid?: false,
+            done?: false,
+            cancelled?: false,
+            client_name: nil,
+            client_size: nil,
+            client_type: nil,
+            client_last_modified: nil
+
+  @type t :: %__MODULE__{
+          progress: integer(),
+          upload_config: String.t() | :atom,
+          upload_ref: String.t(),
+          ref: String.t() | nil,
+          valid?: boolean(),
+          done?: boolean(),
+          cancelled?: boolean(),
+          client_name: String.t() | nil,
+          client_size: integer() | nil,
+          client_type: String.t() | nil,
+          client_last_modified: integer() | nil
+        }
+
+  @doc false
+  def put_progress(%UploadEntry{} = entry, 100) do
+    %UploadEntry{entry | progress: 100, done?: true}
```

Comment:
Pretty minor, but we could do this in a single function head:

`    %UploadEntry{entry | progress: 100, done?: progress == 100}`